openadapt/ml/openai_finetune.py

```diff
@@ -0,0 +1,121 @@
+from openadapt.crud import *
+from openadapt import models, utils
+from copy import deepcopy
+from loguru import logger
+import json
+import openai
+import fire
+
+
```

Comment:
This PR is indeed really useful to me.  👍 

I would like to share my thoughts on the global structure:

IMHO, we want the fine-tuning to happen systematically for each and every model we have. Because a model without fine-tuning is kinda useless, it does not even provides completion as a valid action. Hence, I think we need to make this fine-tuning as generic as possible, not only specific to OpenAI.

I suggest:

1. Wrap everything in a class like this:

```python

class DataProcessor:
    def _retrieve_recording(self, recording_id: str) -> Recording:
        """
        Retrieve the recording from the database
        """
        return crud.get_recording_by_id(recording_id)

    def _condense_data(self, recording: Recording) -> any:
        """
        Condense the action events to remove any unnecessary information"""
        # [TODO] Implement data condenser and sanitizer here, as it is common for all models
        return

    def generate_data(self, recording_id: str) -> any:
        recording = self._retrieve_recording(recording_id)
        condensed_data = self._condense_data(recording)
        return condensed_data


class FineTuner:
    def __init__(self) -> None:
        self.data_processor = DataProcessor()

    def prepare_data_for_tuning(self, file_path: str) -> str:
        """
        Prepare the data for fine-tuning the model. The data should be
        stored in correct format in the given file path.

        The output data is stored in a temporary file and the path to the
        file is returned.
        """
        pass

    def tune_model(self, fine_tune_data_path: Optional[str] = None)->any:
        """
        Fine tune the model with the given data.
        """
        pass

    def tune_model_autoregressively(self, recording_id: str):
        """
        Fine tune the model autoregressively on the given recording.
        """
        data = self.data_processor.generate_data(recording_id)
        # TODO: Save condensed data to a temporary file for fine-tuning
        # TODO: Fine-tune the model with the data
        prep_data_path = self.prepare_data_for_tuning("tmp/data/prompt.json")
        self.tune_model(prep_data_path)
        # TODO: maybe return the fine-tuned model name or fine tuned params


```


Then, in the case of `OpenAITuning`, this piece of code will become:

```python

class OpenAIFineTuner(FineTuner):
    def __init__(self, model_name) -> None:
        super().__init__()
        self.model_name = model_name

    def prepare_data_for_tuning(self, file_path: str) -> str:
        """
        Prepare the data for fine-tuning the model. The data should be
        stored in correct format in the given file path.

        The output data is stored in a temporary file and the path to the
        file is returned.
        """
        output = subprocess.run(
            ["openai", "tools", "fine_tunes.prepare_data", "-f", file_path],
            capture_output=True,
        )
        return output.stdout.decode("utf-8")
    
    def tune_model(self, fine_tune_data_path: Optional[str] = None)->any:
        output = subprocess.run(
            ["openai", "api", "fine_tunes.create", "-t", fine_tune_data_path, "-m", self.model_name],
            capture_output=True,
        )
        return output.stdout.decode("utf-8")

```


The advantage is that tomorrow, when someone wants to fine tune a random model on HuggingFace for example, they can reuse this code here and avoid re-writing the whole boilerplate parts. Also, we enforce that, though the model may be different, the structure of the fine tuning pipelines remain almost identical. This is a huge bonus for this project as we are more or less model agnostic.

Reply:
WDYT ? @FFFiend @abrichr

Reply:
This looks great, I wasn't aware of the ```subprocess``` library 😄

And then for the model api, #379 we can have a FineTuner class attribute for each model.

Reply:
Exactly. In #379, each `CompletionProvider` can have its own `fine_tuner` object. IMHO, this helps a lot in structuring the code and make adoption of new model really easy. Something like

```python

class GPT3CompletionProvider:
    def __init__(self):
        self.fine_tuner = OpenAIFineTuner()


```

Reply:
My idea is that, because we have to deal with many things in parallel (fine tune, completion, reinforced learning...etc), it is better to structure things separately.

Reply:
This looks great, thank you @LaPetiteSouris !

I suggest also looking into the following libraries to see if we can re-use some of their logic or primitives:
- https://github.com/CambioML/pykoi 
- https://github.com/sybrenjansen/mpire
- https://www.ray.io/